Files/Filesystem/StorageEnumerators/UniversalStorageEnumerator.cs

```diff
@@ -0,0 +1,268 @@
+﻿using ByteSizeLib;
+using Files.Extensions;
+using Files.Views.LayoutModes;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Windows.Storage;
+using Windows.Storage.FileProperties;
+using Windows.UI.Xaml.Media.Imaging;
+
+namespace Files.Filesystem.StorageEnumerators
+{
+    public static class UniversalStorageEnumerator
+    {
+        public static async Task<List<ListedItem>> ListEntries(
+            StorageFolder rootFolder,
+            StorageFolderWithPath currentStorageFolder,
+            string returnformat,
+            Type sourcePageType,
+            CancellationToken cancellationToken,
+            Func<List<ListedItem>, Task> intermediateAction
+        )
+        {
+            var tempList = new List<ListedItem>();
+            uint count = 0;
+            var firstRound = true;
+            while (true)
+            {
+                IReadOnlyList<IStorageItem> items;
+                uint maxItemsToRetrieve = 300;
+
+                if (intermediateAction == null)
+                {
+                    // without intermediate action increase batches significantly
+                    maxItemsToRetrieve = 1000;
+                }
+                else if (firstRound)
+                {
+                    maxItemsToRetrieve = 32;
+                    firstRound = false;
+                }
+                try
+                {
+                    items = await rootFolder.GetItemsAsync(count, maxItemsToRetrieve);
+                    if (items == null || items.Count == 0)
+                    {
+                        break;
+                    }
+                }
+                catch (NotImplementedException)
+                {
+                    break;
+                }
+                catch (Exception ex) when (
+                    ex is UnauthorizedAccessException
+                    || ex is FileNotFoundException
+                    || (uint)ex.HResult == 0x80070490) // ERROR_NOT_FOUND
+                {
+                    // If some unexpected exception is thrown - enumerate this folder file by file - just to be sure
+                    items = await EnumerateFileByFile(rootFolder, count, maxItemsToRetrieve);
```

Comment:
@gave92 As per our discussion - i'm using batch file loading and falling back to file by file only in case of exception here